packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrixAndCombine_def.hpp

```diff
@@ -124,21 +125,17 @@ unpack_crs_matrix_row(typename PackTraits<GO, Device>::output_array_type& gids_o
   }
   bool unpack_pids = pids_out.size() > 0;
 
-  const LO num_ent_LO = static_cast<LO>(num_ent); // packValueCount wants this
   const size_t num_ent_beg = offset;
-  const size_t num_ent_len = PackTraits<LO, Device>::packValueCount(num_ent_LO);
+  const size_t num_ent_len = PackTraits<LO, DT>::packValueCount(LO(0));
 
-  const GO gid = 0; // packValueCount wants this
   const size_t gids_beg = num_ent_beg + num_ent_len;
-  const size_t gids_len = num_ent * PackTraits<GO, Device>::packValueCount(gid);
+  const size_t gids_len = num_ent * PackTraits<GO, DT>::packValueCount(GO(0));
 
-  const int pid = 0; // packValueCount wants this
   const size_t pids_beg = gids_beg + gids_len;
-  const size_t pids_len = (unpack_pids) ? num_ent * PackTraits<int, Device>::packValueCount(pid) : 0;
+  const size_t pids_len = (unpack_pids) ? num_ent * PackTraits<int, DT>::packValueCount(int(0)) : 0;
 
   if (num_bytes_per_value == 0) {
-    ST val; // packValueCount wants this
-    num_bytes_per_value = PackTraits<ST, Device>::packValueCount(val);
+    num_bytes_per_value = PackTraits<ST, DT>::packValueCount(ST());
```

Comment:
This will break Stokhos, due to the assumption that default-constructed ST has the right size, but it was broken before anyway ;-) .  Thus, it's OK to defer fixing this to a separate PR.

Reply:
I think this branch is unused and can be removed.  All callers to this function send in a value > 0

Reply:
@tjfulle What's actually the point of the `num_bytes_per_value` branch?

Reply:
I believe when I put the branch in I was combining the several unpacking schemes and this branch allowed for one case that the `num_bytes_per_value` was not known and Scalar was default constructed.  On further combining/refactoring, that case went away.

Reply:
@tjfulle It looks like the idea is that Stokhos is supposed to pass in the right `num_bytes_per_value` value -- is that right?  (@etphipp)

Reply:
Stohkos never calls this function, it calls packValueCount and friends.  This function is a combination of similar functions in the unpackCrsMatrix and unpackWithOwningPIDs functions.  Stokhos use of packValueCount is pretty simple and will allow simplifying packValueCount as @ibaned has advocated.

Reply:
I'm just looking at the code in full now and remember why the branch exists.  When unpacking the `CrsMatrix`, matrix values are not known a-priori, so the only (current) way of getting the size of the values is by the default constructor.  When unpacking in to `CrsArrays`, at least some matrix values are known and those values can be used to determine the size needed to unpack.

Reply:
I've modified the code to push the computation of the number of bytes per value further up stream.  It does not remove the issue, but simplifies this function by removing the branch.